File: src/main/java/eu/xap3y/connectfour/manager/ConfigLoader.java
package eu.xap3y.connectfour.manager;

import eu.xap3y.connectfour.ConnectFour;
import eu.xap3y.connectfour.api.dto.TexterObjDto;
import eu.xap3y.connectfour.api.model.ConfigModel;
import eu.xap3y.connectfour.api.model.PlayerStatModel;
import eu.xap3y.connectfour.service.Texter;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

import java.io.File;
import java.io.IOException;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ConfigLoader {

    public final Map<String, PlayerStatModel> data = new HashMap<>();

    private File dataFile;
    private YamlConfiguration dataConfig;

    public void reload() {
        ConnectFour.getInstance().reloadConfig();
        FileConfiguration cfg = ConnectFour.getInstance().getConfig();

        LangManager.prefix = cfg.getString("prefix", "&8[&6ConnectFour&8] &7");
        ConnectFour.language = cfg.getString("lang", "en");

        ConnectFour.setTexter(new Texter(new TexterObjDto(
                cfg.getString("prefix", "&8[&6ConnectFour&8] &7"),
                false,
                null
        )));

        ConnectFour.setConfigModel(new ConfigModel(
                cfg.getInt("inviteTimeout", 60),
                cfg.getBoolean("tokenFallAnimation", true),
                cfg.getLong("tokenFallSpeed", 120L),
                cfg.getBoolean("inviterStarts", false),
                cfg.getBoolean("winRewards.enabled", false),
                cfg.getStringList("winRewards.commands"),
                cfg.getBoolean("hooks.placeholderAPI", true),
                cfg.getBoolean("hooks.miniPlaceholders", false),
                cfg.getBoolean("metrics", true),
                cfg.getBoolean("updateCheck", true),
                cfg.getBoolean("doubleEscape", true),
                cfg.getInt("moveTimeout", 0)
        ));

        dataFile = new File(ConnectFour.getInstance().getDataFolder(), "data.yml");
        if (!dataFile.exists()) {
            File oldData = new File(ConnectFour.getInstance().getDataFolder(), "storage.json");
            if (oldData.exists()) {
                ConnectFour.getTexter().console("&eOld storage.json file detected, attempting migration to data.yml...");
                boolean migrated = eu.xap3y.connectfour.util.StorageMigrator.migrate(ConnectFour.getInstance(), true, false);
                if (migrated) {
                    ConnectFour.getTexter().console("&aMigration completed successfully!");
                } else {
                    ConnectFour.getTexter().console("&cMigration failed! Please check the console for details.");
                }
            } else {
                ConnectFour.getInstance().saveResource("data.yml", false);
            }
        }

        dataConfig = YamlConfiguration.loadConfiguration(dataFile);
    }

    public void loadData() {
        if (!ConnectFour.getInstance().getDataFolder().exists()) return;

        if (dataConfig.getConfigurationSection("players") == null) {
            return;
        }
        for (String key : dataConfig.getConfigurationSection("players").getKeys(false)) {
            String path = "players." + key + ".";
            PlayerStatModel psm = new PlayerStatModel(
                    key,
                    dataConfig.getString(path + "name", "Unknown"),
                    dataConfig.getInt(path + "gamesPlayed", 0),
                    dataConfig.getInt(path + "wins", 0),
                    dataConfig.getInt(path + "losses", 0),
                    dataConfig.getInt(path + "draws", 0)
            );
            data.put(key, psm);
        }
    }

    /*public void saveData() {
        File dataFile = new File(ConnectFour.getInstance().getDataFolder(), "storage.json");

        try (FileWriter writer = new FileWriter(dataFile, false)) {
            StorageModel storage = new StorageModel(new ArrayList<>(data.values()));
            prettyGson.toJson(storage, writer);
        } catch (Exception ignored) {
        }
    }*/

    public void savePlayerData(Player p0) {
        PlayerStatModel psm = data.get(p0.getUniqueId().toString());
        String path = "players." + p0.getUniqueId() + ".";
        dataConfig.set(path + "name", p0.getName());
        dataConfig.set(path + "gamesPlayed", psm.getGamesPlayed());
        dataConfig.set(path + "wins", psm.getWins());
        dataConfig.set(path + "losses", psm.getLosses());
        dataConfig.set(path + "draws", psm.getDraws());
        try {
            dataConfig.save(dataFile);
        } catch (IOException e) {
            ConnectFour.getTexter().console("&cCould not save player data for " + p0.getName() + "!");
        }

    }

    public void checkPlayer(Player player) {
        String key = player.getUniqueId().toString();
        if (!data.containsKey(key)) {
            data.put(key, new PlayerStatModel(
                    player.getUniqueId().toString(),
                    player.getName(),
                    0,
                    0,
                    0,
                    0
            ));
        }
    }

    public PlayerStatModel getPlayerStats(OfflinePlayer player) {
        return data.get(player.getUniqueId().toString());
    }

    public List<PlayerStatModel> getLeaderboard() {
        int limit = LangManager.getInt("leaderboard.entries", 10);
        return data.values()
                .stream()
                .sorted(Comparator.comparingInt(PlayerStatModel::getWins).reversed())
                .limit(limit)
                .toList();
    }
}

File: core/src/main/java/eu/xap3y/connectfour/utils/ConfigLoader.kt
package eu.xap3y.connectfour.utils

import eu.xap3y.connectfour.ConnectFour
import eu.xap3y.connectfour.models.ConfigModel
import eu.xap3y.connectfour.models.PlayerStatModel
import eu.xap3y.connectfour.models.StorageModel
import eu.xap3y.xalib.managers.Texter
import eu.xap3y.xalib.objects.TexterObj
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.io.File
import kotlin.collections.HashMap

class ConfigLoader(private val plugin: ConnectFour) {

    val data: HashMap<String, PlayerStatModel> = hashMapOf()

    fun reload(){
        plugin.configManager.reloadConfig()
        plugin.prefix = plugin.config.getString("prefix", "&8[&6ConnectFour&8] &7") ?: "&8[&6ConnectFour&8] &7"
        plugin.texter = Texter(
            TexterObj(
                plugin.prefix,
                false,
                null
            )
        )

        plugin.configModel = ConfigModel(
            plugin.config.getInt("inviteTimeout", 60),
            plugin.config.getBoolean("tokenFallAnimation", true),
            plugin.config.getLong("tokenFallSpeed", 120L),
            plugin.config.getBoolean("inviterStarts", false),
            plugin.config.getBoolean("winRewards.enabled", false),
            plugin.config.getStringList("winRewards.commands"),
            plugin.config.getBoolean("hooks.placeholderAPI", true),
            plugin.config.getBoolean("hooks.miniPlaceholders", false),
            plugin.config.getBoolean("metrics", true),
            updates = plugin.config.getBoolean("updateCheck", true)
        )
    }

    fun loadData() {
        if (!plugin.dataFolder.exists()) return
        val dataFile = File(plugin.dataFolder, "storage.json")
        if (!dataFile.exists())
            plugin.saveResource("storage.json", false)

        val json = dataFile.readText()
        try {
            val parsed = Json.decodeFromString<StorageModel>(json)
            parsed.players.forEach {
                data[it.uuid] = it
            }
        } catch (e: Exception) {
            saveData()
        }

    }

    @OptIn(ExperimentalSerializationApi::class)
    fun saveData() {
        val dataFile = File(plugin.dataFolder, "storage.json")

        val prettyJson = Json {
            prettyPrint = true
            prettyPrintIndent = " "
        }

        val json: String = prettyJson.encodeToString(StorageModel(data.values.toList()))
        dataFile.writeText(json)
    }

    fun checkPlayer(player: Player) {
        if (!data.containsKey(player.uniqueId.toString())) {
            data[player.uniqueId.toString()] = PlayerStatModel(player.uniqueId.toString(), player.name, 0, 0, 0, 0)
        }
    }

    fun getPlayerStats(player: OfflinePlayer): PlayerStatModel? {
        //checkPlayer(player)
        return data[player.uniqueId.toString()]
    }

    fun getLeaderboard(): List<PlayerStatModel> {
        return data.values.sortedByDescending { it.wins }.take(10)

    }
}